src/modm/driver/motion/adns9800.hpp

```diff
@@ -9,95 +9,426 @@
+  };
+  MODM_FLAGS8(LaserControl);
+
+  enum class ConfigurationII : uint8_t {
+    Cpi_Reporting_Mode = Bit2,
+    Fixed_FrameRate = Bit3,
+    Disable_AGC = Bit4,
+    Rest_Enable = Bit5,
+    Force_Rest0 = Bit6,
+    Force_Rest1 = Bit7,
+  };
+  MODM_FLAGS8(ConfigurationII);
+
+  // Adns9800 takes periods up to 65535
+  using PeriodType = uint16_t;
+  // The internal timer is running at 50MHz, hence the ratio:
+  using Duration = std::chrono::duration<PeriodType, std::ratio<1, 50_MHz>>;
+
+  // forward declarations
+  struct Data;
+  struct DataAndFaildetect;
+  struct DataAndFaildetectAndMonitoring;
 };
 
-/// @ingroup modm_driver_adns9800
-template < typename Spi, typename Cs >
-class Adns9800 : public adns9800
-{
+/**
+ * ADNS9800 Laser Motion Sensor
+ *
+ * @tparam SpiMaster
+ * @tparam Cs
+ *
+ * @author	Thomas Sommer
+ * @ingroup	modm_driver_adns9800
+ */
+template <class SpiMaster, class Cs>
+class Adns9800 : public adns9800, public modm::SpiDevice<SpiMaster> {
 public:
-	static bool
-	initialise();
+  static constexpr size_t FrameSize = 30 * 30; // Size of CMOS
+  /**
+   * @brief                ShutterConfig boundaries which may be selected by the automatic frame rate control.
+   *                       Periods are expressed as ticks of the device, running at 50MHz.
+   *                       In "constant Framerate mode" period_max determines the framerate.
+   *
+   * @param period_min     Frame period min bound -> max allowed fps
+   * @param period_max     Frame period max bound -> min allowed fps.
+   * @param exposure_max   Exposure period max bound -> max exposure time
+   *
+   * @ingroup modm_driver_adns9800
+   */
+  struct ShutterConfig {
+    PeriodType period_min, period_max, exposure_max;
 
-	static bool
-	isNewMotionDataAvailable();
+    // The datasheet refers to "wait for one frame" without better specification.
+    Duration getOneFrameTime() const {
+      return Duration(exposure_max);
+    }
 
-	static void
-	getDeltaXY(int16_t &delta_x, int16_t &delta_y);
+    /// @cond
+    // Always validate configuration before write: There's no internal protection.
+    // Malformed periods may result in undefined behaviour.
+    bool isValid() {
+      return modm_assert_continue_ignore(period_max >= period_min + exposure_max, "adns9800.shutter",
+      "Invalid shutter configuration: !(period_max >= period_min + exposure_max)");
+    }
+    /// @endcond
+  };
 
-protected:
-	static uint8_t
-	readReg(Register const reg);
+  /**
+   * @brief      Default Initialization routine
+   */
+  void
+  initialize() {
+    powerUp();
+    if(verifyIdentity()) {
+      writeFirmware();
+      laserEnable();
+    }
+  }
 
-	static void
-	writeReg(Register const reg, uint8_t const data);
+  /**
+   * @brief      powerUp() resets the device's internal state after power-loss or user invoked shutdown().
+   */
+	void
+  powerUp() {
+    /// @see power-up sequence, datasheet page 20
+    Cs::reset();
+    modm::this_fiber::sleep_for(10ns);
```

Comment:
Use `modm::delay(10ns)`. The fiber sleep is only implemented for microseconds, since a context switch takes around 1us. (for comparison one cycle at 100MHz = 10ns<sup>-1</sup>.) (even on 550MHz STM32H7 it takes ~250ns).

Reply:
Thought of just some pointers to rotate on context switches but that sounds like there's more 😅 Assembler is not (yet) in my repertoire but I immediately wanna digg deeper into fibers gearwork now.

- 
Wouldn't it be nice to inject this knowledge into the code? 

F.e. an assert in `sleep_for(...)` that checks for the minimum appropriate duration. With the MCUs clock by hand, this min. could even be scaled as needed!

Reply:
Yes, I agree those details should be all taken care of by the function itself. 
Have to think about this some more, for example, on AVRs, a context switch takes 10us, so if you sleep for just 1us in a loop, it'll never yield…

(The [context switch code is here](https://github.com/modm-io/modm/blob/develop/src/modm/processing/fiber/context_arm_m.cpp.in#L15-L68) btw)